src/core/net/netif.cpp

```diff
@@ -135,25 +135,17 @@ otError Netif::SubscribeAllRoutersMulticast(void)
 {
     otError error = OT_ERROR_NONE;
 
-    if (mMulticastAddresses == &kLinkLocalAllNodesMulticastAddress)
```

Comment:
In `Mle()` constructor, `mLinkLocalAllThreadNodes` and `mRealmLocalAllThreadNodes` would be subscribed in `SetMeshLocalPrefix()` and added to the head of `mMulticastAddresses` linked-list. So when comes to`SubscribeAllRoutersMulticast()`, the list would be

> mRealmLocalAllThreadNodes->mLinkLocalAllThreadNodes-> LinkLocalAll-> RealmLocalAll-> RealmLocalAllMplForwarders -> NULL

And after `SubscribeAllRoutersMulticast()`, the list would be

>  mRealmLocalAllThreadNodes->mLinkLocalAllThreadNodes->LinkLocalAllRouters -> RealmLocalAllRouters -> LinkLocalAll -> RealmLocalAll -> RealmLocalAllMplForwarders -> NULL

So when no additional external multicast address subscription, the head of `mMulticastAddresses` list should always be `mRealmLocalAllThreadNodes`.

Reply:
Thanks for explanation. Got it! It's because in our model we will always have `mLinkLocalAllThread` and `mRealLocalAllThread` addresses added from `Mle()` constructor so that head of the linked list will not point to one the constant addresses.

Reply:
Thinking more about this, I think it may be better if we keep the logic as is. 

- I agree that in current way we add the addresses, we would not go through the code path where the `if()` checks can be true.
- But considering that the methods `SubscribeAllRoutersMulticast()` and `UnsubscribeAllRoutersMulticast()` are part a generic `Netif` class, I feel it'd better (and more correct behavior) for them to still check and handle these cases (where the head of list matches one of the constant `AllNodes` or `AllRouters` addresses).
- This way the implementation does not require/assume that other addresses would be in list. And `Netif` class remains more flexible and independent of `Mle`.

What do you think?

Reply:
@abtink . Thanks for your explanation. It sounds reasonable.
will close this PR. Thanks.